src/obj/export.c

```diff
@@ -0,0 +1,653 @@
+{
+    OSSL_PARAM params[MAX_KEY_ATTRS + 1] = { 0 };
+    CK_ATTRIBUTE_TYPE types[MAX_KEY_ATTRS];
+    CK_ATTRIBUTE *type;
+    CK_ATTRIBUTE *value;
+    const unsigned char *val;
+    X509 *x509 = NULL;
+    EVP_PKEY *pkey;
+    int attrnum = 0;
+    int ret;
+    CK_RV rv;
+
+    /* if CKA_CERTIFICATE_TYPE is not present assume CKC_X_509 */
+    type = p11prov_obj_get_attr(crt, CKA_CERTIFICATE_TYPE);
+    if (type) {
+        CK_CERTIFICATE_TYPE crt_type = CKC_X_509;
+        if (type->ulValueLen != sizeof(CK_CERTIFICATE_TYPE)
+            || memcmp(type->pValue, &crt_type, type->ulValueLen) != 0) {
+            return CKR_OBJECT_HANDLE_INVALID;
+        }
+    }
+
+    value = p11prov_obj_get_attr(crt, CKA_VALUE);
+    if (!value) {
+        return CKR_GENERAL_ERROR;
+    }
+
+    val = value->pValue;
+    x509 = d2i_X509(NULL, &val, value->ulValueLen);
+    if (!x509) {
+        return CKR_GENERAL_ERROR;
+    }
+
+    pkey = X509_get0_pubkey(x509);
+    if (!pkey) {
+        rv = CKR_GENERAL_ERROR;
+        goto done;
+    }
+
+    if (EVP_PKEY_is_a(pkey, "RSA")) {
+        for (int i = 0; i < num; i++) {
+            switch (attrs[i].type) {
+            case CKA_MODULUS:
+                types[attrnum] = CKA_MODULUS;
+                params[attrnum] =
+                    OSSL_PARAM_construct_BN(OSSL_PKEY_PARAM_RSA_N, NULL, 0);
+                attrnum++;
+                break;
+            case CKA_PUBLIC_EXPONENT:
+                types[attrnum] = CKA_PUBLIC_EXPONENT;
+                params[attrnum] =
+                    OSSL_PARAM_construct_BN(OSSL_PKEY_PARAM_RSA_E, NULL, 0);
+                attrnum++;
+                break;
+            }
+        }
+    } else if (EVP_PKEY_is_a(pkey, "EC")) {
+        for (int i = 0; i < num; i++) {
+            switch (attrs[i].type) {
+            case CKA_P11PROV_CURVE_NAME:
+                types[attrnum] = CKA_P11PROV_CURVE_NAME;
+                params[attrnum] = OSSL_PARAM_construct_utf8_string(
+                    OSSL_PKEY_PARAM_GROUP_NAME, NULL, 0);
+                attrnum++;
+                break;
+            case CKA_P11PROV_PUB_KEY:
+                types[attrnum] = CKA_P11PROV_PUB_KEY;
+                params[attrnum] = OSSL_PARAM_construct_octet_string(
+                    OSSL_PKEY_PARAM_PUB_KEY, NULL, 0);
+                attrnum++;
+                break;
+            }
+        }
+    } else if (EVP_PKEY_is_a(pkey, ED25519) || EVP_PKEY_is_a(pkey, ED448)) {
+        for (int i = 0; i < num; i++) {
+            switch (attrs[i].type) {
+            case CKA_P11PROV_PUB_KEY:
+                types[attrnum] = CKA_P11PROV_PUB_KEY;
+                params[attrnum] = OSSL_PARAM_construct_octet_string(
+                    OSSL_PKEY_PARAM_PUB_KEY, NULL, 0);
+                attrnum++;
+                break;
+            }
+        }
+    } else {
+        rv = CKR_OBJECT_HANDLE_INVALID;
+        goto done;
+    }
```

Comment:
Out of scope of this PR, but here we should handle the ML-DSA and SLH-DSA certificates here when ready. What tests are using this code path? I though we have already the ML-DSA support for certificates and export already.